src/emit.cpp

```diff
@@ -0,0 +1,76 @@
+#include "emit.h"
+
+#include <map>
+
+#include "debug.h"
+#include "generic_factory.h"
+
+template <>
+const emit_id string_id<emit>::NULL_ID( "null" );
+
+static std::map<emit_id, emit> emits_all;
+
+template<>
+bool string_id<emit>::is_valid() const
+{
+    const auto found = emits_all.find( *this );
+    if( found == emits_all.end() ) {
+        return false;
+    }
+    return found->second.field() != fd_null;
+}
+
+template<>
+const emit &string_id<emit>::obj() const
+{
+    const auto found = emits_all.find( *this );
+    if( found == emits_all.end() ) {
+        debugmsg( "Tried to get invalid emission data: %s", c_str() );
```

Comment:
You could get rid of the boilerplate (and save some lines) by using `generic_factory` for routine purposes.